src/test/fuzz/versionbits.cpp

```diff
@@ -0,0 +1,343 @@
+
+    // between genesis and 2100-01-01
+    const int64_t block_start_time = fuzzed_data_provider.ConsumeIntegralInRange<uint32_t>(1231006505, 4102444800);
+
+    // too many blocks at 10min each might cause uint32_t time to overflow if
+    // block_start_time is at the end of the range above
+    assert(n_blocks < 320000);
+
+    // what values for version will we use to signal / not signal?
+    int32_t ver_signal = fuzzed_data_provider.ConsumeIntegral<int32_t>();
+    int32_t ver_nosignal = fuzzed_data_provider.ConsumeIntegral<int32_t>();
+
+    // select deployment parameters: bit, start time, timeout
+    int bit = fuzzed_data_provider.ConsumeIntegralInRange<int>(0, VERSIONBITS_NUM_BITS-1);
+
+    bool always_active_test = false;
+    bool never_active_test = false;
+    int64_t start_time;
+    int64_t timeout;
+    if (fuzzed_data_provider.ConsumeBool()) {
+        start_time = block_start_time + start_block*600;
+        timeout = block_start_time + end_block*600;
+
+        assert(start_time <= timeout);
+
+        // allow for times to not exactly match a block
+        if (fuzzed_data_provider.ConsumeBool()) start_time += 300;
+        if (fuzzed_data_provider.ConsumeBool()) timeout += 300;
+
+        // this may make timeout too early; if so, don't run the test
+        if (start_time > timeout) return;
+    } else {
+        if (fuzzed_data_provider.ConsumeBool()) {
+            start_time = Consensus::BIP9Deployment::ALWAYS_ACTIVE;
+            timeout = Consensus::BIP9Deployment::NO_TIMEOUT;
+            always_active_test = true;
+        } else {
+            start_time = 1199145601; // January 1, 2008
+            timeout = 1230767999; // December 31, 2008
+            never_active_test = true;
+        }
+    }
+
+    TestConditionChecker checker(start_time, timeout, period, threshold, bit);
+
+    // Early exit if the versions don't signal sensibly for the deployment
+    if (!checker.Condition(ver_signal)) return;
+    if (checker.Condition(ver_nosignal)) return;
+    if (ver_nosignal < 0) return;
+
+    // TOP_BITS should ensure version will be positive
+    assert(ver_signal > 0);
+
+    // Now that we have chosen time and versions, setup to mine blocks
+    Blocks blocks(block_start_time, ver_signal, ver_nosignal);
+
+    /* Strategy:
+     *  * we will mine a final period worth of blocks, with
+     *    randomised signalling according to a mask
+     *  * but before we mine those blocks, we will mine some
+     *    randomised number of prior periods; with either all
+     *    or no blocks in the period signalling
+     *
+     * We establish the mask first, then consume "bools" until
+     * we run out of fuzz data to work out how many prior periods
+     * there are and which ones will signal.
+     */
+
+    // establish the mask
+    uint32_t signalling_mask = fuzzed_data_provider.ConsumeIntegral<uint32_t>();
```

Comment:
Move this declaration down to where it's used first.

Reply:
I'm deliberately getting the mask beforehand so that adding bytes at the end of the fuzz data just adds blocks and doesn't change what gets interpreted as a mask and what gets interpreted as a bool

Reply:
ah! Makes sense. Thanks.